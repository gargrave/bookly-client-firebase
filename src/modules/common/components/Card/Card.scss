@import '../../../../styles/_shadows';

$textColor: #6e6969;
$titleColor: #5d5656;

.card {
  @include boxShadowMed;
  background-color: #f9f9f9;
  border: solid 1px #bbb;
  border-radius: 5px;
  margin: 8px auto;
  max-width: 550px;
  padding: 10px 20px;
}

.hoverableCard:hover {
  background: #e4effc;
  cursor: pointer;
}

.header,
.title,
.text {
  margin: 0;
}

.header,
.title {
  color: $titleColor;
  font-weight: bold;
}

.header {
  font-size: 1.4rem;
}

.detailCard {
  padding: 15px;

  .title {
    font-size: 1.3rem;
    line-height: 2.5rem;
  }

  .title + .cardText {
    font-size: 1.1rem;
  }

  .cardText + .cardText {
    margin-top: 8px;
  }
}

.listDetailCard {
  padding: 10px
}
