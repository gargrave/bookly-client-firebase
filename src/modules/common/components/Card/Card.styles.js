import styled from 'react-emotion'
import { colors, shadows } from '../../../../styles/'

export const StyledWrapper = styled('div')`
  ${shadows.medium};

  background-color: ${colors.cardBg};
  border: solid 1px ${colors.cardBorder};
  border-radius: 5px;
  margin: 8px auto;
  max-width: 550px;
  padding: 10px 20px;

  &.hoverable:hover {
    background: ${colors.cardHoveredBg};
    cursor: pointer;
  }
`
