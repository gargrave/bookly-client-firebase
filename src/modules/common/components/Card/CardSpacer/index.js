import CardSpacer from './CardSpacer'

export default CardSpacer
