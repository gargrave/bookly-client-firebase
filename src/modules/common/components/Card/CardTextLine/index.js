import CardTextLine from './CardTextLine'

export default CardTextLine
