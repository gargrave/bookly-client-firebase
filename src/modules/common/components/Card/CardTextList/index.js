import CardTextList from './CardTextList'

export default CardTextList
