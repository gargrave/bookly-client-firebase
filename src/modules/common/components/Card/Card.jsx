// @flow
import React from 'react'
import { any, bool, func } from 'prop-types'
import styled from 'react-emotion'

import { buildClass } from '../../../../utils/cssHelpers'

import CardHeader from './CardHeader'
import CardSpacer from './CardSpacer'
import CardTextLine from './CardTextLine'
import CardTextList from './CardTextList'

import { colors, shadows } from '../../../../styles/'

type Props = {
  children?: any,
  hoverable?: boolean,
  onClick?: Function,
}

const Styled = styled('div')`
  ${shadows.medium};

  background-color: ${colors.cardBg};
  border: solid 1px ${colors.cardBorder};
  border-radius: 5px;
  margin: 8px auto;
  max-width: 550px;
  padding: 10px 20px;
  text-align: center;

  &.hoverable:hover {
    background: ${colors.cardHoveredBg};
    cursor: pointer;
  }
`

const Card = ({ children, hoverable = false, onClick }: Props) => {
  return (
    <Styled className={buildClass({ hoverable })} onClick={onClick}>
      {children}
    </Styled>
  )
}

Card.Header = CardHeader
Card.Spacer = CardSpacer
Card.TextLine = CardTextLine
Card.TextList = CardTextList

Card.propTypes = {
  children: any,
  hoverable: bool,
  onClick: func,
}

export default Card
