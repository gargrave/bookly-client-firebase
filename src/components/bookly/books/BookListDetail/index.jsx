// @flow
import React from 'react';
import { bool, func, shape, string } from 'prop-types';

import type { Book } from '../../../../constants/flowtypes';

import Card from '../../../common/Card';
import CardTextLine from '../../../common/Card/CardTextLine';

type Props = {
  book: Book,
  onClick: Function,
  showAuthor?: boolean,
};

const BookListDetail = ({
  book,
  onClick,
  showAuthor,
}: Props) => {
  const {
    author,
    title,
  } = book;
  const authorName = `${author.firstName} ${author.lastName}`;

  return (
    <Card
      classes={[
        'list-detail-card',
        'book-list-detail-card',
      ]}
      onClick={onClick}
    >
      <CardTextLine bold text={title} />
      {showAuthor && <CardTextLine text={authorName} />}
    </Card>
  );
};

BookListDetail.propTypes = {
  book: shape({
    author: shape({
      firstName: string.isRequired,
      lastName: string.isRequired,
    }),
    title: string.isRequired,
  }).isRequired,
  onClick: func.isRequired,
  showAuthor: bool,
};

export default BookListDetail;
