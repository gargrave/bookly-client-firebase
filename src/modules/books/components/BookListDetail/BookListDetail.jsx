// @flow
import React from 'react';
import { bool, func, shape, string } from 'prop-types';

import type { Book } from '../../flowtypes';

import Card from '../../../common/components/Card/Card';

type Props = {
  book: Book,
  onClick: Function,
  showAuthor?: boolean,
};

const BookListDetail = ({
  book,
  onClick,
  showAuthor,
}: Props) => {
  const { author, title } = book;
  const authorName = `${author.firstName} ${author.lastName}`;

  return (
    <Card
      classes={[
        'list-detail-card',
        'book-list-detail-card',
      ]}
      onClick={onClick}
    >
      <Card.TextLine bold text={title} />
      {showAuthor && <Card.TextLine text={authorName} />}
    </Card>
  );
};

BookListDetail.propTypes = {
  book: shape({
    author: shape({
      firstName: string.isRequired,
      lastName: string.isRequired,
    }),
    title: string.isRequired,
  }).isRequired,
  onClick: func.isRequired,
  showAuthor: bool,
};

export default BookListDetail;
