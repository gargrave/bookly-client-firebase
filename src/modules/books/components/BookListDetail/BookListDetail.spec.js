import { bookMocks } from '../../../../globals/mocks/'
import { ComponentBuilder } from '../../../../utils/testHelpers'

import Card from '../../../common/components/Card/Card'

import BookListDetail from './BookListDetail'

const testBook = bookMocks[0]

const defaultProps = {
  book: testBook,
  onClick: jest.fn(),
}

const authorName = book => `${book.author.firstName} ${book.author.lastName}`

const builder = new ComponentBuilder(BookListDetail, defaultProps)

describe('BookListDetail', () => {
  let component

  beforeEach(() => {
    component = builder.shallowGetComponent()
  })

  describe('basic rendering', () => {
    test('matches the snapshot', () => {
      expect(component).toMatchSnapshot()
    })

    test('renders one Card', () => {
      expect(component.find(Card)).toHaveLength(1)
    })
  })

  describe('title text display', () => {
    test('renders the title text', () => {
      const textLines = component.find(Card.TextLine)
      const titleText = component.find({ text: defaultProps.book.title })
      expect(textLines).toHaveLength(1)
      expect(titleText).toHaveLength(1)
    })
  })

  describe('sorting string', () => {
    const sortBy = 'sort by this'

    test('does not render a sorting string if there is none', () => {
      const textLines = component.find(Card.TextLine)
      const sortByText = component.find({ text: sortBy })
      expect(textLines).toHaveLength(1)
      expect(sortByText).toHaveLength(0)
    })

    test('does not render a sorting string if there is none', () => {
      const bookWithSort = { ...testBook, sortBy }
      component = builder.shallowGetComponent({ book: bookWithSort })
      const textLines = component.find(Card.TextLine)
      const sortByText = component.find({ text: `Sorted by: ${sortBy}` })
      expect(textLines).toHaveLength(2)
      expect(sortByText).toHaveLength(1)
    })
  })

  describe('author name display', () => {
    test('does not render author name by default', () => {
      const textLines = component.find(Card.TextLine)
      const authorText = textLines.find({ text: authorName(defaultProps.book) })
      expect(textLines).toHaveLength(1)
      expect(authorText).toHaveLength(0)
    })

    test('renders the author when "showAuthor" is true', () => {
      component = builder.shallowGetComponent({ showAuthor: true })
      const textLines = component.find(Card.TextLine)
      const authorText = textLines.find({ text: authorName(defaultProps.book) })
      expect(textLines).toHaveLength(2)
      expect(authorText).toHaveLength(1)
    })
  })
})
