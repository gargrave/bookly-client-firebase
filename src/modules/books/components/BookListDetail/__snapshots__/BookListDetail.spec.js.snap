// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookListDetail basic rendering matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookListDetail
    book={
      Object {
        "author": Object {
          "bookCount": 1,
          "created": 2017-07-01T01:16:43.881Z,
          "firstName": "Aldous",
          "id": "oiw223h3qndlksadffl3ksdf",
          "lastName": "Huxley",
          "updated": 2017-07-01T01:16:43.881Z,
        },
        "created": 2018-01-14T01:36:26.220Z,
        "id": "sv3hmbefpc95MRXQsTHr",
        "sortBy": "",
        "title": "Brave New World",
        "updated": 2018-01-14T01:36:26.220Z,
      }
    }
    onClick={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <CardTextLine
          bold={true}
          text="Brave New World"
        />,
        false,
        "",
      ],
      "hoverable": true,
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "bold": true,
          "text": "Brave New World",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
      "",
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <CardTextLine
            bold={true}
            text="Brave New World"
          />,
          false,
          "",
        ],
        "hoverable": true,
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "bold": true,
            "text": "Brave New World",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        "",
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
